### ISSUE TITLE
High cpu load on memory increase

### ISSUE THREAD

**Issue by dodoma** [cesanta/mjs#25]
CPU load almost 100% on mjs manage about 3M memory.

Here is my test code and data, tested on osx x86_64.

load.c
```c
#include "mjs.h"
#include <dlfcn.h>
#include <string.h>
#include <sys/time.h>

int main(int argc, char *argv[])
{
    /*
     * prepare big object
     */
    char strbuf[10000];
    int len = 0;
    FILE *fp = fopen("./x.json", "r");
    if (fp) {
        memset(strbuf, 0x0, sizeof(strbuf));
        len = fread(strbuf, 1, 10000, fp);
        fclose(fp);
        printf("read %d bytes from x.json\n", len);
    } else {
        printf("can't open x.json\n");
        return 1;
    }


    struct mjs *jsm = mjs_create();

    mjs_set_ffi_resolver(jsm, dlsym);

    mjs_err_t jerr;
    mjs_val_t res;
    jerr = mjs_exec_file(jsm, "./x.js", &res);
    if (jerr != MJS_OK) {
        printf("%s\n", mjs_strerror(jsm, jerr));
    }

    mjs_val_t global = mjs_get_global(jsm);
    mjs_val_t object = mjs_get(jsm, global, "myObject", ~0);
    mjs_val_t function = mjs_get(jsm, object, "test_func", ~0);

    struct timeval total_tvs, total_tve;
    gettimeofday(&total_tvs, NULL);

    for (int i = 0; i < 100; i++) {
        struct timeval tvs, tve;
        gettimeofday(&tvs, NULL);

        mjs_val_t a1 = mjs_mk_string(jsm, strbuf, len, 0);

        jerr = mjs_call(jsm, &res, function, object, 1, a1);
        if (jerr != MJS_OK) {
            printf("%s", mjs_strerror(jsm, jerr));
        }

        gettimeofday(&tve, NULL);

        //unsigned long ela = (tve.tv_sec - tvs.tv_sec) * 1000000ul + (tve.tv_usec - tvs.tv_usec);
        printf("call %d elapsed %d seconds\n", i, tve.tv_sec - tvs.tv_sec);
    }

    gettimeofday(&total_tve, NULL);

    printf("totoal elapsed %d seconds\n", total_tve.tv_sec - total_tvs.tv_sec);

    mjs_destroy(jsm);

    return 0;
}
```

x.js
```js
let myObject = {
    arrs: [],
    test_func: function(orders) {
        this.arrs[this.arrs.length] = JSON.parse(orders);
        //JSON.parse(orders);
        //print(orders);
        print(this.arrs.length);
    }
};
```

x.json
```js
{"orders":[{"type_desc":"已买入","count":2000,"price":460000,"type":4,"bid":true,"channel":"CXA","code":"OCC","id":"36C8E003"},{"type_desc":"已买入","count":2000,"price":320000,"type":4,"bid":true,"channel":"CXA","code":"DRM","id":"37CBE003"},{"type_desc":"已买入","count":2000,"price":500000,"type":4,"bid":true,"channel":"CXA","code":"AWE","id":"10CDE003"},{"type_desc":"已卖出","count":2000,"price":325000,"type":5,"bid":false,"channel":"CXA","code":"DRM","id":"20CEE003"},{"type_desc":"委托买入","count":2000,"price":250000,"type":1,"bid":true,"channel":"CXA","code":"BLG","id":"31D0E003"},{"type_desc":"委托买入","count":2000,"price":440000,"type":1,"bid":true,"channel":"CXA","code":"BIG","id":"57D0E003"},{"type_desc":"取消委托","count":2000,"price":465000,"type":3,"bid":false,"channel":"CXA","code":"OCC","id":"80D0E003"},{"type_desc":"已买入","count":10000,"price":95000,"type":4,"bid":true,"channel":"CXA","code":"FGR","id":"B6D0E003"},{"type_desc":"已卖出","count":2000,"price":460000,"type":5,"bid":false,"channel":"CXA","code":"OCC","id":"97D1E003"},{"type_desc":"已买入","count":2000,"price":320000,"type":4,"bid":true,"channel":"CXA","code":"DRM","id":"FBD1E003"},{"type_desc":"已卖出","count":324,"price":325000,"type":5,"bid":false,"channel":"CXA","code":"DRM","id":"8CD2E003"},{"type_desc":"已卖出","count":1676,"price":325000,"type":5,"bid":false,"channel":"CXA","code":"DRM","id":"2ED3E003"},{"type_desc":"已买入","count":2000,"price":320000,"type":4,"bid":true,"channel":"CXA","code":"DRM","id":"CED3E003"},{"type_desc":"已买入","count":2000,"price":450000,"type":4,"bid":true,"channel":"CXA","code":"OCC","id":"DDD3E003"},{"type_desc":"已买入","count":2000,"price":455000,"type":4,"bid":true,"channel":"CXA","code":"OCC","id":"F2D4E003"},{"type_desc":"已卖出","count":2000,"price":460000,"type":5,"bid":false,"channel":"CXA","code":"OCC","id":"40D6E003"},{"type_desc":"已卖出","count":2000,"price":455000,"type":5,"bid":false,"channel":"ASXT","code":"OCC","id":"44D7E003"},{"type_desc":"委托买入","count":1250,"price":450000,"type":1,"bid":true,"channel":"CXA","code":"OCC","id":"4AD7E003"},{"type_desc":"委托买入","count":2000,"price":310000,"type":1,"bid":true,"channel":"CXA","code":"LBT","id":"9AD8E003"},{"type_desc":"委托买入","count":2000,"price":415000,"type":1,"bid":true,"channel":"CXA","code":"PLS","id":"A7D8E003"},{"type_desc":"已买入","count":2000,"price":685000,"type":4,"bid":true,"channel":"CXA","code":"MMJ","id":"AADAE003"},{"type_desc":"已买入","count":2000,"price":2570000,"type":4,"bid":true,"channel":"CXA","code":"WAM","id":"83DCE003"},{"type_desc":"已买入","count":2000,"price":165000,"type":4,"bid":true,"channel":"CXA","code":"SCU","id":"4EE1E003"},{"type_desc":"取消委托","count":2000,"price":805000,"type":3,"bid":true,"channel":"CXA","code":"AC8","id":"75E1E003"},{"type_desc":"已买入","count":2000,"price":800000,"type":4,"bid":true,"channel":"CXA","code":"AC8","id":"96E3E003"},{"type_desc":"取消委托","count":2000,"price":690000,"type":3,"bid":false,"channel":"CXA","code":"MMJ","id":"A2E4E003"},{"type_desc":"已买入","count":2000,"price":675000,"type":4,"bid":true,"channel":"CXA","code":"MMJ","id":"E9E4E003"},{"type_desc":"已卖出","count":2000,"price":685000,"type":5,"bid":false,"channel":"CXA","code":"MMJ","id":"6DE5E003"},{"type_desc":"取消委托","count":7400,"price":96000,"type":3,"bid":false,"channel":"CXA","code":"FGR","id":"1FF4E003"},{"type_desc":"已卖出","count":2000,"price":805000,"type":5,"bid":false,"channel":"CXA","code":"AC8","id":"4DF4E003"},{"type_desc":"已买入","count":2000,"price":800000,"type":4,"bid":true,"channel":"CXA","code":"AC8","id":"A7FDE003"},{"type_desc":"已买入","count":2000,"price":680000,"type":4,"bid":true,"channel":"CXA","code":"MMJ","id":"6FFEE003"},{"type_desc":"取消委托","count":2000,"price":775000,"type":3,"bid":true,"channel":"CXA","code":"AC8","id":"ABFFE003"},{"type_desc":"委托买入","count":2000,"price":770000,"type":1,"bid":true,"channel":"CXA","code":"AC8","id":"D8FFE003"},{"type_desc":"已卖出","count":2000,"price":790000,"type":5,"bid":false,"channel":"CXA","code":"AC8","id":"E5FFE003"},{"type_desc":"取消委托","count":2000,"price":685000,"type":3,"bid":false,"channel":"CXA","code":"MMJ","id":"AD00E103"},{"type_desc":"已卖出","count":2000,"price":680000,"type":5,"bid":false,"channel":"CXA","code":"MMJ","id":"E300E103"},{"type_desc":"取消委托","count":2000,"price":680000,"type":3,"bid":false,"channel":"CXA","code":"MMJ","id":"F104E103"},{"type_desc":"已买入","count":500,"price":25000,"type":4,"bid":true,"channel":"ASXT","code":"AGO","id":"1D05E103"},{"type_desc":"取消委托","count":500,"price":26000,"type":3,"bid":false,"channel":"CXA","code":"AGO","id":"2405E103"},{"type_desc":"取消委托","count":10000,"price":96000,"type":3,"bid":false,"channel":"CXA","code":"FGR","id":"2B05E103"},{"type_desc":"取消委托","count":500,"price":25000,"type":3,"bid":false,"channel":"CXA","code":"AGO","id":"F805E103"},{"type_desc":"已买入","count":2000,"price":780000,"type":4,"bid":true,"channel":"CXA","code":"AC8","id":"190BE103"},{"type_desc":"取消委托","count":2000,"price":170000,"type":3,"bid":false,"channel":"CXA","code":"SCU","id":"840CE103"},{"type_desc":"取消委托","count":2000,"price":502500,"type":3,"bid":false,"channel":"CXAM","code":"AWE","id":"FE0FE103"},{"type_desc":"已买入","count":3000,"price":495000,"type":4,"bid":true,"channel":"CXA","code":"AWE","id":"EF10E103"},{"type_desc":"取消委托","count":2000,"price":500000,"type":3,"bid":true,"channel":"CXA","code":"AWE","id":"FC10E103"},{"type_desc":"已卖出","count":1254,"price":505000,"type":5,"bid":false,"channel":"CXA","code":"AWE","id":"1012E103"},{"type_desc":"已卖出","count":2000,"price":165000,"type":5,"bid":false,"channel":"CXA","code":"SCU","id":"C915E103"},{"type_desc":"已买入","count":2000,"price":405000,"type":4,"bid":true,"channel":"CXA","code":"ESE","id":"1217E103"},{"type_desc":"取消委托","count":2000,"price":410000,"type":3,"bid":false,"channel":"CXA","code":"ESE","id":"A117E103"},{"type_desc":"取消委托","count":2000,"price":385000,"type":3,"bid":true,"channel":"CXA","code":"ESE","id":"AA17E103"},{"type_desc":"取消委托","count":2000,"price":395000,"type":3,"bid":false,"channel":"CXA","code":"ESE","id":"D317E103"},{"type_desc":"取消委托","count":1494,"price":410000,"type":3,"bid":false,"channel":"CXA","code":"ESE","id":"EA17E103"},{"type_desc":"取消委托","count":1494,"price":405000,"type":3,"bid":false,"channel":"ASXT","code":"ESE","id":"F117E103"},{"type_desc":"取消委托","count":1494,"price":395000,"type":3,"bid":false,"channel":"CXA","code":"ESE","id":"F717E103"},{"type_desc":"取消委托","count":1494,"price":390000,"type":3,"bid":false,"channel":"CXA","code":"ESE","id":"0618E103"},{"type_desc":"取消委托","count":1494,"price":405000,"type":3,"bid":false,"channel":"ASXT","code":"ESE","id":"2218E103"},{"type_desc":"取消委托","count":1494,"price":390000,"type":3,"bid":false,"channel":"CXA","code":"ESE","id":"3018E103"},{"type_desc":"已卖出","count":1494,"price":395000,"type":5,"bid":false,"channel":"ASXT","code":"ESE","id":"3E18E103"},{"type_desc":"已买入","count":2000,"price":400000,"type":4,"bid":true,"channel":"ASXT","code":"ESE","id":"FC20E103"},{"type_desc":"已买入","count":2000,"price":775000,"type":4,"bid":true,"channel":"ASXT","code":"CPH","id":"D621E103"},{"type_desc":"取消委托","count":2000,"price":770000,"type":3,"bid":true,"channel":"CXA","code":"CPH","id":"D821E103"},{"type_desc":"已卖出","count":2000,"price":780000,"type":5,"bid":false,"channel":"CXA","code":"CPH","id":"4322E103"},{"type_desc":"委托卖出","count":10000,"price":95000,"type":2,"bid":false,"channel":"CXA","code":"FGR","id":"9C22E103"},{"type_desc":"取消委托","count":2000,"price":775000,"type":3,"bid":true,"channel":"CXA","code":"CPH","id":"2824E103"},{"type_desc":"取消委托","count":2000,"price":2575000,"type":3,"bid":false,"channel":"CXAM","code":"WAM","id":"7E25E103"},{"type_desc":"已卖出","count":2000,"price":2570000,"type":5,"bid":false,"channel":"CXA","code":"WAM","id":"0926E103"},{"type_desc":"已卖出","count":2000,"price":785000,"type":5,"bid":false,"channel":"CXA","code":"AC8","id":"7426E103"},{"type_desc":"委托卖出","count":500,"price":24000,"type":2,"bid":false,"channel":"CXA","code":"AGO","id":"3127E103"},{"type_desc":"已卖出","count":2000,"price":325000,"type":5,"bid":false,"channel":"CXA","code":"DRM","id":"B42BE103"},{"type_desc":"未知类型","count":10000,"price":990000,"type":0,"bid":true,"channel":"CXA","code":"PAI","id":"212DE103"},{"type_desc":"取消委托","count":2000,"price":675000,"type":3,"bid":false,"channel":"CXA","code":"MMJ","id":"C12EE103"},{"type_desc":"取消委托","count":2000,"price":670000,"type":3,"bid":false,"channel":"CXA","code":"MMJ","id":"B72FE103"},{"type_desc":"已卖出","count":2000,"price":665000,"type":5,"bid":false,"channel":"CXA","code":"MMJ","id":"0431E103"},{"type_desc":"委托买入","count":2000,"price":660000,"type":1,"bid":true,"channel":"CXA","code":"MMJ","id":"BE31E103"},{"type_desc":"委托买入","count":2000,"price":320000,"type":1,"bid":true,"channel":"CXA","code":"DRM","id":"2233E103"},{"type_desc":"委托卖出","count":3000,"price":497500,"type":2,"bid":false,"channel":"CXAM","code":"AWE","id":"FD33E103"},{"type_desc":"委托买入","count":2000,"price":765000,"type":1,"bid":true,"channel":"ASXT","code":"CPH","id":"5236E103"},{"type_desc":"委托卖出","count":2000,"price":405000,"type":2,"bid":false,"channel":"ASXT","code":"ESE","id":"8636E103"}]}
```

Here is the output:

```
read 9705 bytes from x.json
1 
call 0 elapsed 0 seconds
2 
call 1 elapsed 0 seconds
3 
call 2 elapsed 0 seconds
4 
call 3 elapsed 0 seconds
5 
call 4 elapsed 0 seconds
6 
call 5 elapsed 0 seconds
7 
call 6 elapsed 0 seconds
8 
call 7 elapsed 0 seconds
9 
call 8 elapsed 0 seconds
10 
call 9 elapsed 0 seconds
11 
call 10 elapsed 0 seconds
12 
call 11 elapsed 0 seconds
13 
call 12 elapsed 0 seconds
14 
call 13 elapsed 0 seconds
15 
call 14 elapsed 0 seconds
16 
call 15 elapsed 1 seconds
17 
call 16 elapsed 0 seconds
18 
call 17 elapsed 0 seconds
19 
call 18 elapsed 0 seconds
20 
call 19 elapsed 1 seconds
21 
call 20 elapsed 0 seconds
22 
call 21 elapsed 0 seconds
23 
call 22 elapsed 1 seconds
24 
call 23 elapsed 0 seconds
25 
call 24 elapsed 0 seconds
26 
call 25 elapsed 0 seconds
27 
call 26 elapsed 1 seconds
28 
call 27 elapsed 1 seconds
29 
call 28 elapsed 1 seconds
30 
call 29 elapsed 1 seconds
31 
call 30 elapsed 1 seconds
32 
call 31 elapsed 1 seconds
33 
call 32 elapsed 1 seconds
34 
call 33 elapsed 1 seconds
35 
call 34 elapsed 0 seconds
36 
call 35 elapsed 1 seconds
37 
call 36 elapsed 1 seconds
38 
call 37 elapsed 1 seconds
39 
call 38 elapsed 1 seconds
40 
call 39 elapsed 1 seconds
41 
call 40 elapsed 1 seconds
42 
call 41 elapsed 4 seconds
43 
call 42 elapsed 5 seconds
44 
call 43 elapsed 3 seconds
45 
call 44 elapsed 4 seconds
46 
call 45 elapsed 3 seconds
47 
call 46 elapsed 2 seconds
48 
call 47 elapsed 3 seconds
49 
call 48 elapsed 1 seconds
50 
call 49 elapsed 1 seconds
51 
call 50 elapsed 1 seconds
52 
call 51 elapsed 2 seconds
53 
call 52 elapsed 4 seconds
54 
call 53 elapsed 2 seconds
55 
call 54 elapsed 1 seconds
56 
call 55 elapsed 1 seconds
57 
call 56 elapsed 3 seconds
58 
call 57 elapsed 3 seconds
59 
call 58 elapsed 3 seconds
60 
call 59 elapsed 1 seconds
61 
call 60 elapsed 2 seconds
62 
call 61 elapsed 4 seconds
63 
call 62 elapsed 10 seconds
64 
call 63 elapsed 7 seconds
65 
call 64 elapsed 8 seconds
66 
call 65 elapsed 8 seconds
67 
call 66 elapsed 9 seconds
68 
call 67 elapsed 13 seconds
69 
call 68 elapsed 7 seconds
70 
call 69 elapsed 2 seconds
71 
call 70 elapsed 2 seconds
72 
call 71 elapsed 4 seconds
73 
call 72 elapsed 6 seconds
74 
call 73 elapsed 4 seconds
75 
call 74 elapsed 2 seconds
76 
call 75 elapsed 3 seconds
77 
call 76 elapsed 4 seconds
78 
call 77 elapsed 7 seconds
79 
call 78 elapsed 5 seconds
80 
call 79 elapsed 2 seconds
81 
call 80 elapsed 3 seconds
82 
call 81 elapsed 5 seconds
83 
call 82 elapsed 8 seconds
84 
call 83 elapsed 6 seconds
85 
call 84 elapsed 2 seconds
86 
call 85 elapsed 3 seconds
87 
call 86 elapsed 6 seconds
88 
call 87 elapsed 9 seconds
89 
call 88 elapsed 6 seconds
90 
call 89 elapsed 3 seconds
91 
call 90 elapsed 4 seconds
92 
call 91 elapsed 6 seconds
93 
call 92 elapsed 14 seconds
94 
call 93 elapsed 17 seconds
95 
call 94 elapsed 19 seconds
96 
call 95 elapsed 18 seconds
97 
call 96 elapsed 20 seconds
98 
call 97 elapsed 25 seconds
99 
call 98 elapsed 20 seconds
100 
call 99 elapsed 20 seconds
totoal elapsed 387 seconds
```

My load:
```
PID   COMMAND      %CPU TIME     #TH  #WQ  #POR MEM    PURG CMPR PGRP PPID STATE          BOOSTS     %CPU_ME %CPU_OTHRS UID  FAULT COW  MSGS MSGR SYSB SYSM CSW    PAGE IDLE POWE
4750  load         98.8 01:27.67 1/1  0    10   2240K  0B   0B   4750 978  running *0[1]      0.00000 0.00000    501  1310  126  36   15   110  63   72816+ 0    0    98.8
```

**Comment by cpq** [comment]
@dimonomid could you take a look please? Maybe GC is interfering. The execution time seems extremely uneven, and indeed shows a big bump starting from certain point.

**Comment by dodoma** [comment]
It's gc to eat the cpu load.

I add one line at the bottom of mjs_create()
```c
  mjs->inhibit_gc = 1;
```
these codes will be done in 1 second.

**Comment by dodoma** [comment]
Suggest do gc periodly(by time or memory size), If it's hard to short the gc check time.

**Comment by dimonomid** [comment]
TL;DR:

Try to do the following: find this line:
```c
#define _MJS_STRING_BUF_RESERVE 100
```
and change 100 to, say, 10000.

And then compile with these flags:
```
-DMJS_DEFAULT_OBJECT_ARENA_SIZE=500 -DMJS_DEFAULT_PROPERTY_ARENA_SIZE=500
```
-------
GC is invoked when preallocated arenas for objects and properties (and buffer for strings) are about to be exhausted.

By default the arenas for object/properties contain 20 objects, and when it's not enough (and GC can't release anything), arena size is incremented by 10.

When strings buffer is exhausted, its size is incremented by 100.

These defaults are probably fine for the embedded environment (which is our primary target), but not so fine for the machine with lots of memory available.

Currently, only initial size of object/property arenas can be tweaked (with the compile flags I mentioned above), other params are hardcoded. We'll need to fix this of course, but for now, could you please perform the modifications I mentioned, and see if it helps?

**Comment by dodoma** [comment]
Sounds good, I'll try it tomorrow. And maybe some other questions :)

**Comment by dodoma** [comment]
dimonomid, I tried you advice just now(thanks for the detail explain), I suggest to do following modify on mjs gc:

```c
 MJS_PRIVATE int gc_strings_is_gc_needed(struct mjs *mjs) {
   struct mbuf *m = &mjs->owned_strings;
-  return (double) m->len / (double) m->size > 0.9;
+  if ((double)(m->size - m->len) / (double)_MJS_STRING_BUF_RESERVE < 0.1) {
+      mbuf_resize(m, m->size + _MJS_STRING_BUF_RESERVE);
+      return 1;
+  }
+  return 0;
 }
```

It can avoid unnecessary mjs_gc() on:
1. mjs->owned_strings.len / mjs->owned_strings.size `always` greater than 0.9, if mjs->owned_strings.len grows very big after mjs long time running.

2.  If we mark mjs->need_gc = 1; after gc_strings_is_gc_needed(), we'd better grow up owned_string.size right now, to avoid repeat mjs_gc() call if gc_compact_strings() can't do very usefull compact.

**Comment by dodoma** [comment]
By the way, I seted both _MJS_STRING_BUF_RESERVE, MJS_DEFAULT_OBJECT_ARENA_SIZE, and MJS_DEFAULT_PROPERTY_ARENA_SIZE to 1048576. And seted arena grow up size to 524288.

This runs fine on my 4GB environment, maybe it will be usefull If we offer some `non-embed` environment documents.

**Comment by cpq** [comment]
Can we close this?

**Comment by dodoma** [comment]
@dimonomid do what ever you want, it's ok now for me.

**Comment by dimonomid** [comment]
Before closing this, I want to make these parameters easily adjustable with the compile flags (without the need to edit the code). Also, this is a relevant point:

> mjs->owned_strings.len / mjs->owned_strings.size always greater than 0.9, if mjs->owned_strings.len grows very big after mjs long time running.

Not sure we'll address it now, but let me think more.

**Comment by dodoma** [comment]
En, let's consider the extrimily situation. we push the 'x.josn' into an array for 10000 times. we will reproduce the problem.

**Comment by cpq** [comment]
@dodoma mJS focuses on embedded environment. Supporting those corner cases is not our priority.

**Comment by dimonomid** [comment]
These memory parameters are now adjustable with compiler flags: https://github.com/cesanta/mjs/commit/3ccc378af6177ace11c8a2b94b129b4127d807e0

Closing this one.
